controllers/storagecluster/cephrbdmirrors.go

```diff
@@ -0,0 +1,136 @@
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return fmt.Errorf("Unable to retrieve CephRbdMirror %v: %v", cephRbdMirror.Name, err)
+			}
+			continue
+		}
+		if cephRbdMirror.GetDeletionTimestamp().IsZero() {
+			r.Log.Info("Deleting CephRbdMirror.", "CephRbdMirror", klog.KRef(cephRbdMirror.Namespace, cephRbdMirror.Name))
+			err := r.Client.Delete(context.TODO(), existing)
+			if err != nil {
+				r.Log.Info("Retrying deletion of CephRbdMirror.", "CephRbdMirror", klog.KRef(existing.Namespace, existing.Name))
+				err := r.Client.Delete(context.TODO(), existing)
+				if err != nil {
+					r.Log.Error(err, "Failed to delete CephRbdMirror.", "CephRbdMirror", klog.KRef(existing.Namespace, existing.Name))
+					return fmt.Errorf("Failed to delete CephRbdMirror %v: %v", existing.Name, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// newCephRbdMirrorInstances returns the cephRbdMirror instances that should be created
+// on first run.
+func (r *StorageClusterReconciler) newCephRbdMirrorInstances(initData *ocsv1.StorageCluster) ([]*cephv1.CephRBDMirror, error) {
+	ret := []*cephv1.CephRBDMirror{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      generateNameForCephRbdMirror(initData),
+				Namespace: initData.Namespace,
+			},
+			Spec: cephv1.RBDMirroringSpec{
+				Count:     1,
+				Resources: defaults.GetDaemonResources("rbd-mirror", initData.Spec.Resources),
+			},
+		},
+	}
+	for _, obj := range ret {
+		err := controllerutil.SetControllerReference(initData, obj, r.Scheme)
+		if err != nil {
+			r.Log.Error(err, "Unable to set controller reference for CephRBDMirror.", "CephRBDMirror", klog.KRef(obj.Namespace, obj.Name))
+			return nil, err
+		}
+	}
+	return ret, nil
+}
+
+// ensureCreated ensures that cephRbdMirror resources exist in the desired state.
+func (obj *ocsCephRbdMirrors) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+	reconcileStrategy := ReconcileStrategy(instance.Spec.ManagedResources.CephRbdMirrors.ReconcileStrategy)
+	if reconcileStrategy == ReconcileStrategyIgnore {
+		return nil
+	}
+
+	cephRbdMirrors, err := r.newCephRbdMirrorInstances(instance)
+	if err != nil {
+		return err
+	}
+
+	if !instance.Spec.Mirroring.Enabled {
+		return r.deleteCephRbdMirrorInstance(cephRbdMirrors)
+	}
+
+	for _, cephRbdMirror := range cephRbdMirrors {
+		existing, err := r.fetchCephRbdMirrorInstance(cephRbdMirror)
+
+		if err != nil {
+			if errors.IsNotFound(err) {
+				r.Log.Info("Creating CephRbdMirror.", "CephRbdMirror", klog.KRef(cephRbdMirror.Namespace, cephRbdMirror.Name))
+				err = r.Client.Create(context.TODO(), cephRbdMirror)
+				if err != nil {
+					r.Log.Error(err, "Failed to create CephRbdMirror.", "CephRbdMirror", klog.KRef(cephRbdMirror.Namespace, cephRbdMirror.Name))
+					return err
+				}
+				continue
+			}
+			r.Log.Error(err, "Unable to create CephRbdMirror.", "CephRbdMirror", klog.KRef(cephRbdMirror.Namespace, cephRbdMirror.Name))
```

Comment:
I think this log msg needs to be corrected as your get call failed not the create call.

Reply:
```suggestion
			r.Log.Error(err, "Failed to get CephRbdMirror.", "CephRbdMirror", klog.KRef(cephRbdMirror.Namespace, cephRbdMirror.Name))
```

Reply:
The goal of reconciling here is to create or update the resource (delete is a separate function call), So returning an error for "get" is not inappropriate.

Better I change the error message like:

r.Log.Error(err, "Failed to ensure CephRbdMirror is present.", "CephRbdMirror", klog.KRef(cephRbdMirror.Namespace, cephRbdMirror.Name))

Reply:
means i can't decide to update or delete the CephRbdMirror resource